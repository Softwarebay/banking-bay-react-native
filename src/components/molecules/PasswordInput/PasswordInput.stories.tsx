import { text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react-native';
import { CenterStory } from 'utils/storybook/CenterStory';

import { PasswordInput } from './PasswordInput';

storiesOf('Molecules/PasswordInput', module)
	.addDecorator((getStory) => <CenterStory>{getStory()}</CenterStory>)
	.add('Default', () => (
		<PasswordInput
			placeholder={text('placeholder', 'Enter your password')}
			testID={text('testID', 'PasswordInput-test-id')}
		/>
	));
