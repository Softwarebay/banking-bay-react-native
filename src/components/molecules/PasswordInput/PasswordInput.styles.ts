import { StyleSheet } from 'react-native';

export const styles = StyleSheet.create({
	container: {
		height: 50,
	},
	iconBtn: {
		right: 0,
	},
	inputField: {
		flex: 1,
		marginRight: 20,
	},
});
