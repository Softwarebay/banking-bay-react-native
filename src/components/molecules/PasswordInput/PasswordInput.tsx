import { forwardRef, useState } from 'react';
import { StyleProp, ViewStyle } from 'react-native';

import {
	Card,
	IconButton,
	InputField,
	InputFieldHandle,
} from 'components/atoms';

import { colors } from 'theme';
import { styles } from './PasswordInput.styles';

export interface Props {
	placeholder: string;
	testID: string;
	style?: StyleProp<ViewStyle>;
}

export const PasswordInput = forwardRef<InputFieldHandle, Props>(
	({ placeholder, style, testID }, ref) => {
		const [hidePassword, setHidePassword] = useState(true);
		return (
			<Card style={[styles.container, style]} testID={testID}>
				<InputField
					ref={ref}
					placeholder={placeholder}
					secureTextEntry={hidePassword}
					style={styles.inputField}
					testID={`input-field-${testID}`}
				/>
				<IconButton
					btnStyle={styles.iconBtn}
					iconColor={colors.mainDark}
					iconName="eye-off"
					iconSize={16}
					testID={`icon-btn-${testID}`}
					onPress={() => setHidePassword((hidden) => !hidden)}
				/>
			</Card>
		);
	}
);
