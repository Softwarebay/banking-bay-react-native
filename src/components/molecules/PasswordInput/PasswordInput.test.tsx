import { fireEvent, render, screen } from '@testing-library/react-native';
import testRenderer from 'react-test-renderer';

import { PasswordInput } from './PasswordInput';

import type { Props } from './PasswordInput';

describe('PasswordInput', () => {
	const props: Props = {
		placeholder: 'Enter password',
		testID: 'password-input-test-id',
		style: {
			height: 50,
		},
	};

	it('should match snapshot', () => {
		const tree = testRenderer.create(<PasswordInput {...props} />).toJSON();

		expect(tree).toMatchSnapshot();
	});

	it('should render input field and eye-off icon button', () => {
		render(<PasswordInput {...props} />);

		const passwordInput = screen.getByTestId('password-input-test-id');
		const inputField = screen.getByTestId('input-field-password-input-test-id');
		const iconBtn = screen.getByTestId('icon-btn-password-input-test-id');
		const eyeOffIcon = screen.getByTestId('eye-off-icon-test-id');

		expect(passwordInput.children).toHaveLength(2);
		expect(inputField).not.toBeNull();
		expect(iconBtn).not.toBeNull();
		expect(eyeOffIcon).not.toBeNull();
	});

	it('should apply additional styles', () => {
		render(<PasswordInput {...props} />);

		const passwordInput = screen.getByTestId('password-input-test-id');

		expect(passwordInput).toHaveStyle({
			height: 50,
		});
	});

	it('should initially hide password', () => {
		render(<PasswordInput {...props} />);

		const inputField = screen.getByTestId('input-field-password-input-test-id');

		expect(inputField.props).toHaveProperty('secureTextEntry', true);
	});

	it('should handle type events', () => {
		render(<PasswordInput {...props} />);

		const inputField = screen.getByTestId('input-field-password-input-test-id');

		fireEvent.changeText(inputField, 'password');

		expect(inputField.props).toHaveProperty('value', 'password');
	});

	it('should show/hide password after press on eye-off icon button', () => {
		render(<PasswordInput {...props} />);

		const iconBtn = screen.getByTestId('icon-btn-password-input-test-id');

		fireEvent.press(iconBtn);

		const inputFieldPasswordShown = screen.getByTestId(
			'input-field-password-input-test-id'
		);

		expect(inputFieldPasswordShown.props).toHaveProperty(
			'secureTextEntry',
			false
		);

		fireEvent.press(iconBtn);

		const inputFieldPasswordHidden = screen.getByTestId(
			'input-field-password-input-test-id'
		);

		expect(inputFieldPasswordHidden.props).toHaveProperty(
			'secureTextEntry',
			true
		);
	});
});
