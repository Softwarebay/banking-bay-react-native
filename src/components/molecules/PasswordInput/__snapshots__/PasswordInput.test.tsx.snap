// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
      },
      Object {
        "flexDirection": "row",
      },
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderRadius": 10,
          "shadowColor": "#161966",
          "shadowOffset": Object {
            "height": 6,
            "width": 0,
          },
          "shadowOpacity": 0.08,
          "shadowRadius": 16,
        },
        Array [
          Object {
            "height": 55,
            "paddingHorizontal": 20,
            "paddingVertical": 14,
          },
          Object {
            "height": 50,
          },
        ],
      ],
    ]
  }
  testID="password-input-test-id"
>
  <TextInput
    autoCorrect={false}
    keyboardType="default"
    onChangeText={[Function]}
    placeholder="Enter password"
    secureTextEntry={true}
    style={
      Array [
        Object {
          "color": "#1B1D4D",
          "fontFamily": "MulishRegular",
          "fontSize": 15,
        },
        Object {
          "flex": 1,
          "marginRight": 20,
        },
      ]
    }
    testID="input-field-password-input-test-id"
    value=""
  />
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    hitSlop={20}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
        },
        Object {
          "right": 0,
        },
        false,
      ]
    }
    testID="icon-btn-password-input-test-id"
  >
    <SvgMock
      color="#1B1D4D"
      height={16}
      testID="eye-off-icon-test-id"
      width={16}
    />
  </View>
</View>
`;
